### PR TITLE
[HttpFoundation] Add mime type guessing by file extension

### PR DESCRIPTION
**Pull request by ipf**
This adds mime type guessing by file extension.
Sometimes file information is only stored as metadata, so we don't have access
to a file. So, for guessing the mime type this does it just by using a string (i.e. 'jpg).

| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | no
| License       | MIT

### CODE REVIEW

**Comment by Hanmac** [comment]
i am okay with adding such a "extension" to "mime-type" guesser,
but i would prefer if it can use some system files like that, like the mime.types from apache.

File: src/Symfony/Component/HttpFoundation/File/MimeType/MimeTypeByExtensionGuesser.php [hunk 1]
```diff
@@ -0,0 +1,1019 @@
...
+     * A map of file extensions and their mime types.
+     *
+     * This list has been placed under the public domain by the Apache HTTPD project.
+     * This list has been updated from upstream on 2017-03-13.
+     *
+     * @see http://svn.apache.org/repos/asf/httpd/httpd/trunk/docs/conf/mime.types
+     *
+     * @var array
+     */
+    protected $defaultExtensions = array(
```
Comment (xabbuh) [inline src/Symfony/Component/HttpFoundation/File/MimeType/MimeTypeByExtensionGuesser.php]: `private`

File: src/Symfony/Component/HttpFoundation/File/MimeType/MimeTypeByExtensionGuesser.php [hunk 2]
```diff
@@ -0,0 +1,1019 @@
...
+        'avi' => 'video/x-msvideo',
+        'movie' => 'video/x-sgi-movie',
+        'smv' => 'video/x-smv',
+        'ice' => 'x-conference/x-cooltalk',
+    );
+
+    /**
+     * {@inheritdoc}
+     */
+    public function guess($extension)
```
Comment (xabbuh) [inline src/Symfony/Component/HttpFoundation/File/MimeType/MimeTypeByExtensionGuesser.php]: This is wrong. The interface states that `guess()` must expect the file path and not only its extension.

**Comment by xabbuh** [comment]
What about registering this also as an alternative in `MimeTypeGuesser` that will be used when neither `FileinfoMimeTypeGuesser` nor `FileBinaryMimeTypeGuesser` yield a result?

**Comment by xabbuh** [comment]
Alternatively, instead of inlining the map of extensions to mime types inside the class, we could include the original resources file and allow to pass the file to be used in the constructor (defaulting to the file that is shipped with the HttpFoundation component).

**Comment by ipf** [comment]
Hi @xabbuh, thanks a lot for your reviews. Rebuilding the `MimeTypeGuesser` is definetly a nice thing, but rebuilding all that is quite a lot of work.
Using the original file is also a great idea, but parsing it on each call may be too much overhead instead of my small and simple solution. What do you think?

**Comment by xabbuh** [comment]
> but rebuilding all that is quite a lot of work

What do you mean with "rebuilding all that"?

> Using the original file is also a great idea, but parsing it on each call may be too much overhead instead of my small and simple solution.

This should indeed not happen inside the constructor to avoid overhead for requests that don't make use of the mime type guesser or for cases where one of the other guesser returned a result, but only when the extension based guesser actually does have to perform some work. I don't expect the parsing to be too heavy in these cases. But you may prove me wrong and we should then indeed think about some caching.

File: src/Symfony/Component/HttpFoundation/File/MimeType/MimeTypeByExtensionGuesser.php [hunk 3]
```diff
@@ -0,0 +1,1023 @@
...
+    );
+
+    /**
+     * @param string $extension The extension of a file (i.e. 'jpg' or 'uvp')
+     *
+     * @return string|null
+     */
+    public function guess($extension)
+    {
+        return isset($this->defaultExtensions[$extension]) ? $this->defaultExtensions[$extension] : null;
```
Comment (dosten) [inline src/Symfony/Component/HttpFoundation/File/MimeType/MimeTypeByExtensionGuesser.php]: I don't know if it worth it, but PNG, JPEG, etc are valid extensions, what about lowercase the extension before use it?

File: src/Symfony/Component/HttpFoundation/Tests/File/MimeType/MimeTypeByExtensionGuesserTest.php [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+    /**
+     * @dataProvider extensionDataProvider
+     *
+     * @param $extension
+     * @param $mimeType
+     */
+    public function testMimeTypeGuessing($extension, $mimeType)
+    {
+        $result = $this->guesser->guess($extension);
```
Comment (dmaicher) [inline src/Symfony/Component/HttpFoundation/Tests/File/MimeType/MimeTypeByExtensionGuesserTest.php]: could be a bit shorter like this? Imho no setup method or private member needed here?

```php
$result = (new MimeTypeByExtensionGuesser())->guess($extension);
```

File: src/Symfony/Component/HttpFoundation/Tests/File/MimeType/MimeTypeByExtensionGuesserTest.php [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+    public function setUp()
+    {
+        $this->guesser = new MimeTypeByExtensionGuesser();
+    }
+
+    /**
+     * @dataProvider extensionDataProvider
+     *
+     * @param $extension
```
Comment (dmaicher) [inline src/Symfony/Component/HttpFoundation/Tests/File/MimeType/MimeTypeByExtensionGuesserTest.php]: useless without any type info

**Comment by nicolas-grekas** [comment]
I agree with @xabbuh about the list: we dont want to maintain it, so we definitely should build it from a reference source. This could happen with a cache warmer I guess.

**Comment by nicolas-grekas** [comment]
Moving to 4.1. Rebase on master might be needed, where PHP 7.1 features can be used btw.

**Comment by fabpot** [comment]
The list of extensions is already part of the `MimeTypeExtensionGuesser` class, so I would extract it from there so that we can reuse it easily. Anyway, any interest in finishing this one? Or should it be closed?

**Comment by teohhanhui** [comment]
This may not be the right place to raise the issue, but the DX around MIME type guessing is kind of broken. See https://github.com/symfony/symfony/issues/15460#issuecomment-380138425 for example.

**Comment by fabpot** [comment]
Closing in favor of #29896

**Pull request closed without merging**